File: parent_node.py
from .node import node
from .gameboard import N_COLS, N_ROWS

class parent_node(node):
    def __init__(self, board, computed_boards, depth = 7):
        self.n_children = len(board.valid_moves)
        self.board = board
        self.children = []
        self.path = []
        self.antinode = False
        self.depth = depth + (self.n_children - 7)
        self.computed_boards = computed_boards
        self.computed_boards.end_count = 0
        self.computed_boards.node_count = 0

        if N_COLS * N_ROWS - sum(self.board._column_heighths) < 15:
            self.depth = 20
        if self.n_children < 5:
            self.depth = 24
        if self.n_children == 5:
            self.depth = 10

    def recommend_move(self):
        self.evaluate()
        max_score = -1
        for child in self.children:
            if child.score > max_score:
                max_score = child.score
                best_move = child.path[0]

        print(self)
        print("computer calculates %i with score %f" % (best_move, max_score))
        return best_move

    def evaluate(self):
        for m in self.board.valid_moves:
            child = node(self, m)
            child.evaluate()

            self.children.append(child)

            # trimming the tree
            if child.score == 0:
                if self.antinode:
                    self.score = 0
                    break

            if child.score == 1:
                if not self.antinode:
                    self.score = 1
                    break

        self.score = max([child.score for child in self.children])

File: src/parent_node.jl
include("node.jl")
include("gameboard.jl")


Base.@kwdef mutable struct ParentNode <: ANode
    board::Gameboard
    children = Node[]
    
    depth = 0
    computed_boards = ComputedBoards()
end


function recommend_move!(p_node::ParentNode, max_nodes=10000)
    max_depth = 3
    add_children!(p_node)

    while p_node.computed_boards.node_count < max_nodes
        println("evaluating at depth $max_depth")
        reevaluate!(p_node, max_depth)

        if 1 >= sum(!child.terminal for child in p_node.children)
            break
        end

        if max_depth >= N_ROWS*N_COLS
            break
        end
            
        max_depth += 1
    end

    max_idx = argmax(child.score for child in p_node.children)

    for child in p_node.children
        s = child.score
        if s > -0.99
            println(child.last_move, "\t", child.score)
        else
            println(child.last_move, "\t x")
        end
    end

    best_move = p_node.children[max_idx].last_move
    max_score = p_node.children[max_idx].score

    println("computer calculates $best_move with score $max_score")
    println("evaluated $(p_node.computed_boards.node_count) boards")
    println("$(p_node.computed_boards.node_links) links")
    return best_move
end



function reevaluate!(p_node::ParentNode, max_depth)
    for child in p_node.children
        reevaluate!(child, max_depth)
    end
end



